cmd/repo-init/api.go

```diff
@@ -0,0 +1,459 @@
+		),
+	))
+	handler := metrics.TraceHandler(simplifier, apiMetrics.HTTPRequestDuration, apiMetrics.HTTPResponseSize)
+	mux := http.NewServeMux()
+	mux.HandleFunc("/api/cluster-profiles", handler(clusterProfileHandler()).ServeHTTP)
+	mux.HandleFunc("/api/configs", handler(configHandler()).ServeHTTP)
+	mux.HandleFunc("/api/config-validations", handler(configValidationHandler()).ServeHTTP)
+	httpServer := &http.Server{Addr: ":" + strconv.Itoa(port), Handler: mux}
+	interrupts.ListenAndServe(httpServer, 5*time.Second)
+	logger.Debug("Ready to serve HTTP requests.")
+}
+
+func clusterProfileHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		disableCORS(w)
+		switch r.Method {
+		case http.MethodGet:
+			marshalled, err := json.Marshal(getClusterProfiles())
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			w.WriteHeader(http.StatusOK)
+			_, err = w.Write(marshalled)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				logrus.WithError(err).Error("unable to marshall response")
+				return
+			}
+		case http.MethodOptions:
+			w.WriteHeader(http.StatusNoContent)
+		default:
+			w.WriteHeader(http.StatusMethodNotAllowed)
+		}
+	}
+}
+
+func configValidationHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		disableCORS(w)
+		switch r.Method {
+		case http.MethodPost:
+			validateConfig(w, r)
+		case http.MethodOptions:
+			w.WriteHeader(http.StatusNoContent)
+		default:
+			w.WriteHeader(http.StatusMethodNotAllowed)
+		}
+	}
+}
+
+func configHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		disableCORS(w)
+		switch r.Method {
+		case http.MethodGet:
+			loadConfigs(w, r, retrieveAndLockAvailable)
+		case http.MethodPost:
+			generateConfig(w, r, retrieveAndLockAvailable)
+		case http.MethodOptions:
+			w.WriteHeader(http.StatusNoContent)
+		default:
+			w.WriteHeader(http.StatusMethodNotAllowed)
+		}
+	}
+}
+
+func disableCORS(w http.ResponseWriter) {
+	if disableCors {
+		w.Header().Set("Access-Control-Allow-Origin", "*")
+		w.Header().Set("Access-Control-Allow-Methods", "*")
+		w.Header().Set("Access-Control-Allow-Headers", "*")
+	}
+}
+
+func loadConfigs(w http.ResponseWriter, r *http.Request, repoGetterFunc RepoGetter) {
+	org := r.URL.Query().Get("org")
+	repo := r.URL.Query().Get("repo")
+
+	if org == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		logrus.Error("no org provided")
+		_, _ = w.Write([]byte("You must provide an org when querying configs."))
+		return
+	}
+
+	githubUser := r.Header.Get("github_user")
+	availableRepo, err := repoGetterFunc(githubUser)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		logrus.WithError(err).Error("unable to get available repo")
+		_, _ = w.Write([]byte("Unable to retrieve a copy of the o/release repo to use. This probably just means that all of them are in use. Please try again in a few seconds."))
```

Comment:
I normally use `Bad Request` as "failed" but not catastrophic failure, but I've definitely been part of many debates in the past for what response codes to use for what. IMO, the http spec response codes have not quite caught up to where REST APIs are today, so it's often necessary to get a little creative. I don't think a 5xx level error code is really right here since it's not a catastrophic failure, and I don't think a 2xx level with an error message is really the right thing to do either since a 2xx implies "all good".